### PR TITLE
Asynchronous LogUnit

### PR DESCRIPTION
**Pull request by Maithem**
## Overview
Requests to the BatchWriter use to block on CompletableFuture.get(),
this change makes both the BatchWriter and LogUnitServer asynchronous.
This is needed to reduce the total number of threads in the server.

Why should this be merged: Reduces the number of threads on the server and improves performance under high write load. 

Related issue(s) (if applicable): #1018

## Checklist (Definition of Done):

- [ ] There are no TODOs left in the code
- [ ] [Coding conventions](https://github.com/CorfuDB/CorfuDB/wiki/Corfu-Style-Guidelines) (e.g. for logging, unit tests) have been followed
- [ ] Change is covered by automated tests
- [ ] Public API has Javadoc

### CODE REVIEW

File: infrastructure/src/main/java/org/corfudb/infrastructure/log/LogMetadata.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+            long currentStreamTail = streamTails.getOrDefault(streamId, Address.NON_ADDRESS);
+            streamTails.put(streamId, Math.max(currentStreamTail, globalTail));
+        }
+
+        // We should also consider checkpoint metadata while updating the tails.
+        // This is important because there could be streams that have checkpoint
+        // data on the checkpoint stream, but not entries on the regular stream.
+        // If those streams are not updated, then clients would observe those
+        // streams as empty, which is not correct.
+        if (entry.hasCheckpointMetadata()) {
```
Comment (xnull) [inline infrastructure/src/main/java/org/corfudb/infrastructure/log/LogMetadata.java]: you can flip this condition, if `hasCheckpointMetadata ` returns false, add a log message and do return.
The same for the next condition `if (Address.isAddress(streamTailAtCP))`.
In the result, you get rid of enclosed if blocks, which reduces cognitive complexity

File: infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogFiles.java [hunk 2]
```diff
@@ -108,16 +118,53 @@ public StreamLogFiles(ServerContext serverContext, boolean noVerify) {
...
+        LogMetadata metadata = new LogMetadata();
+        long start = System.currentTimeMillis();
+        for (long currentSegment = startSegment; currentSegment <= endSegment; currentSegment++) {
+            // TODO(Maithem): factor out getSegmentHandleForAddress to allow getting
+            // segments by segment number
+            SegmentHandle sh = getSegmentHandleForAddress(currentSegment * RECORDS_PER_LOG_FILE + 1);
+            try {
+                for (Map.Entry<Long, AddressMetaData> record : sh.getKnownAddresses().entrySet()) {
+                    // skip trimmed entries
+                    if (record.getKey() < startingAddress) continue;
```
Comment (xnull) [inline infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogFiles.java]: please use curly braces

File: infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogFiles.java [hunk 3]
```diff
@@ -200,15 +247,20 @@ static int getChecksum(int num) {
     }
 
     @Override
-    public long getGlobalTail() {
-        return globalTail.get();
+    public Tails getTails() {
+        Map<UUID, Long> tails = new HashMap<>(logMetadata.getStreamTails().size());
+
+        for (Map.Entry<UUID, Long> entry : logMetadata.getStreamTails().entrySet()) {
+            tails.put(entry.getKey(), new Long(entry.getValue()));
```
Comment (xnull) [inline infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogFiles.java]: `Long.valueOf`

**Comment by Maithem** [comment]
Rebuilding this on Master

**Pull request closed without merging**